custom/icds_reports/utils/data_pull_scripts/ap_data_pull.sql

```diff
@@ -0,0 +1,286 @@
+
+COPY (
+select district_name, block_name, supervisor_name,awc_name,
+SUM(pse_eligible) as pse_eligible,
+SUM(CASE WHEN pse_days_attended is not null and pse_days_attended=0 then 1 ELSE 0 END) as pse_0,
+CASE WHEN SUM(pse_eligible) is not null and SUM(pse_eligible)>0 THEN SUM(CASE WHEN pse_days_attended is not null and pse_days_attended=0 then 1 ELSE 0 END)::float/SUM(pse_eligible)*100 ELSE 0 END as percent_0,
+SUM(CASE WHEN pse_days_attended is not null and pse_days_attended BETWEEN 1 and 7 then 1 ELSE 0 END) as pse_1_7,
+CASE WHEN SUM(pse_eligible) is not null and SUM(pse_eligible)>0 THEN SUM(CASE WHEN pse_days_attended is not null and pse_days_attended BETWEEN 1 and 7 then 1 ELSE 0 END)::float/SUM(pse_eligible)*100 ELSE 0 END as percent_1_7,
+SUM(CASE WHEN pse_days_attended is not null and pse_days_attended BETWEEN 8 and 14 then 1 ELSE 0 END)as  pse_8_14,
+CASE WHEN SUM(pse_eligible) is not null and SUM(pse_eligible)>0 THEN SUM(CASE WHEN pse_days_attended is not null and pse_days_attended BETWEEN 8 and 14 then 1 ELSE 0 END)::float/SUM(pse_eligible)*100 ELSE 0 END as percent_8_14,
+SUM(CASE WHEN pse_days_attended is not null and pse_days_attended between 15 and 21 then 1 ELSE 0 END)as pse_15_21,
+CASE WHEN SUM(pse_eligible) is not null and SUM(pse_eligible)>0 THEN SUM(CASE WHEN pse_days_attended is not null and pse_days_attended between 15 and 21 then 1 ELSE 0 END)::float/SUM(pse_eligible)*100 ELSE 0 END as percent_15_21,
+SUM(CASE WHEN pse_days_attended is not null and pse_days_attended>21 then 1 ELSE 0 END) as pse_gt_21,
+CASE WHEN SUM(pse_eligible) is not null and SUM(pse_eligible)>0 THEN SUM(CASE WHEN pse_days_attended is not null and pse_days_attended>21 then 1 ELSE 0 END)::float/SUM(pse_eligible)*100 ELSE 0 END as percent_gt_21,
+month
+from  awc_location inner join
+"child_health_monthly" child_health on (awc_location.doc_id=child_health.awc_id and
+    awc_location.state_id='f98e91aa003accb7b849a0f18ebd7039' and
+    district_is_test=0 and block_is_test=0 and supervisor_is_test=0 and
+    awc_is_test=0 and child_health.month='2019-10-01'
+    )
+WHERE child_health.age_in_months>36
+group by  district_name, block_name, supervisor_name,awc_name, month
```

Comment:
This grouping has to be done on the master node. The query might be much faster, if you start only by grouping by supervisor_id and then rollup from there (or again can try just doing that in a subselect), but that will push the aggregation work from the master to the nodes.

Reply:
Even just adding supervisor_id to the group by, without changing the query may help there, since that will let it know the whole aggregation can happen on the workers.

Reply:
OK, I will add supervisor_id in the group by thats not a probem. But just to make sure we are on same page, this data pull need per row per awc not per row per supervisor, so I can't just group by supervisor and roll up

Reply:
right, but grouping by `supervisor_id, awc_id` is the same as by `awc_id` in terms of results since there is never a situation where one awc appears under two different supervisors, but it tell citus that it can do that aggregation locally, and doesnt need to bring the data to the master node. like all these suggestions, they are just hunches that need to be verified by `EXPLAIN` but i believe it will lead to a much more efficient query.